vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java

```diff
@@ -398,40 +409,80 @@ public List<SamplesheetHeaderName> retrieveGenomics() {
+    for (int rowId = 1; rowId < sampleRegistrationSheet.getLastRowNum(); rowId++) {
+      Row row = sampleRegistrationSheet.getRow(rowId);
+      List<String> mandatoryInputs = new ArrayList<>();
+      for (SamplesheetHeaderName name : SamplesheetHeaderName.values()) {
+        if(name.isMandatory) {
+          mandatoryInputs.add(SpreadsheetMethods.cellToStringOrNull(row.getCell(
+              header.indexOf(name))));
+        }
+      }
+      // break when cells in row are undefined
+      if (mandatoryInputs.stream().anyMatch(Objects::isNull)) {
+        break;
+      }
+
+      // mandatory not filled in --> invalid
+      if (mandatoryInputs.stream().anyMatch(String::isBlank)) {
+        return Result.fromError(new InvalidSpreadsheetRow(
+            SpreadsheetInvalidationReason.MISSING_INPUT, rowId));
+      }
+
+      String replicateIDInput = SpreadsheetMethods.cellToStringOrNull(row.getCell(
+          header.indexOf(SamplesheetHeaderName.BIOLOGICAL_REPLICATE_ID))).trim();
+      String conditionInput = SpreadsheetMethods.cellToStringOrNull(row.getCell(
+          header.indexOf(SamplesheetHeaderName.CONDITION))).trim();
+      // Sample uniqueness needs to be guaranteed by condition and replicate ID
+      String concatenatedSampleID = replicateIDInput + conditionInput;
+      if (concatenatedSampleIDs.contains(concatenatedSampleID)) {
+        return Result.fromError(new InvalidSpreadsheetRow(
+            SpreadsheetInvalidationReason.DUPLICATE_ID, rowId, replicateIDInput));
+      }
+      concatenatedSampleIDs.add(concatenatedSampleID);
+    }
+    return Result.fromValue(null);
+  }
+
   public List<NGSRowDTO> getFilledRows() {
     List<NGSRowDTO> rows = new ArrayList<>();
-    for (int i = 1; i < sampleRegistrationSheet.getLastRowNum(); i++) {
-      Row row = sampleRegistrationSheet.getRow(i);
-      Cell analysisTypeCell = row.getCell(header.indexOf(SamplesheetHeaderName.SEQ_ANALYSIS_TYPE));
-      Cell sampleLabelCell = row.getCell(header.indexOf(SamplesheetHeaderName.SAMPLE_LABEL));
-      Cell replicateIDCell = row.getCell(
-          header.indexOf(SamplesheetHeaderName.BIOLOGICAL_REPLICATE_ID));
-      Cell conditionCell = row.getCell(header.indexOf(SamplesheetHeaderName.CONDITION));
-      Cell speciesCell = row.getCell(header.indexOf(SamplesheetHeaderName.SPECIES));
-      Cell specimenCell = row.getCell(header.indexOf(SamplesheetHeaderName.SPECIMEN));
-      Cell analyteCell = row.getCell(header.indexOf(SamplesheetHeaderName.ANALYTE));
-      Cell commentCell = row.getCell(header.indexOf(SamplesheetHeaderName.CUSTOMER_COMMENT));
-
-      Supplier<Stream<Cell>> mandatoryCellStreamSupplier = () -> Stream.of(analysisTypeCell,
-          sampleLabelCell,
-          replicateIDCell, conditionCell, speciesCell, specimenCell, analyteCell);
-
-      if (mandatoryCellStreamSupplier.get().anyMatch(Objects::isNull)) {
+
+    for (int rowId = 1; rowId < sampleRegistrationSheet.getLastRowNum(); rowId++) {
```

Comment:
`sampleRegistrationSheet.getLastRowNum()` might not return the last filled row but the last row that was ever touched. Is this handled correctly?

Reply:
Rows are 0-based, is this considered?

Reply:
good catch, it was wrong in three different methods :)